src/jit/codegenlegacy.cpp

```diff
@@ -19462,14 +19462,30 @@ regMaskTP CodeGen::genCodeForCall(GenTreePtr call, bool valUsed)
                         regNumber indCallReg;
 
                         case IAT_VALUE:
+                        {
                             //------------------------------------------------------
                             // Non-virtual direct calls to known addressess
                             //
                             // The vast majority of calls end up here....  Wouldn't
                             // it be nice if they all did!
                             CLANG_FORMAT_COMMENT_ANCHOR;
 #ifdef _TARGET_ARM_
-                            if (!arm_Valid_Imm_For_BL((ssize_t)addr))
+                            // We may use direct call for some of recursive calls
+                            // as we can safely estimate the distance from the call site to the top of the method
+                            const int MAX_PROLOG_SIZE = MAX_USHORT_SIZE_T;
+                            const int codeOffset      = getEmitter()->emitCurCodeOffset + getEmitter()->emitCurIGsize;
+                            if (compiler->gtIsRecursiveCall(call->AsCall()) &&
+                                MAX_PROLOG_SIZE + codeOffset + 4 < 16777216) // if code offset fits in imm24
+                            {
+                                getEmitter()->emitIns_Call(emitter::EC_FUNC_TOKEN, methHnd,
+                                                           INDEBUG_LDISASM_COMMA(sigInfo) NULL, args, retSize,
+                                                           gcInfo.gcVarPtrSetCur, gcInfo.gcRegGCrefSetCur,
+                                                           gcInfo.gcRegByrefSetCur, ilOffset, REG_NA, REG_NA, 0,
+                                                           0,     /* ireg, xreg, xmul, disp */
+                                                           false, /* isJump */
```

Comment:
@adiaaida The code what I tried does not have more than 120 characters per line. The formatter wasn't complaining about it. It first merged two lines into one(I have no idea why) then it complains the line length is over 120. Finally it ends up with splitting the line into two. Like below:

```
[more indentation]    gcInfo.gcRegByrefSetCur, ilOffset, REG_NA, REG_NA, 0,
[more indentation]    0, // ireg, xreg, xmul, disp
```
which is the code of current `master`.